Abort Kubernetes Ingress update if Kubernetes API call fails

Regner opened the pull request:
Currently if a Kubernetes API call fails we potentially remove a working service from Traefik. This changes it so if a Kubernetes API call fails we abort out of the ingress update and use the current working config. Github issue: #1240 

Tests are currently failing but wanted to get the PR up for a few reasons:
1. Confirm this is how we want to solve this issue (abort the whole ingress update if a single API call fails)
2. That there isn't a better way to do the tests
3. I would like the tests to catch a specific error, but don't know what error to expect being raised from the API call

Fixes #1240

timoreimann commented on provider/kubernetes_test.go:
```diff
@@ -1524,11 +1524,80 @@ func TestServiceAnnotations(t *testing.T) {
+		apiEndpointsError: true,
+	}
+	provider = Kubernetes{}
+	_, err = provider.loadIngresses(client)
```
I suggest we convert the test into a [table-driven one](https://github.com/golang/go/wiki/TableDrivenTests) using [sub-tests](https://blog.golang.org/subtests). Among several benefits, this should improve readability and reduce the amount of test code needed.

timoreimann commented on provider/kubernetes_test.go:
```diff
@@ -1524,11 +1524,80 @@ func TestServiceAnnotations(t *testing.T) {
+	_, err := provider.loadIngresses(client)
+
+	if err == nil {
+		t.Fatal("Should have raised error when loading services but did not.")
```
This can be `t.Error` in my view.

timoreimann replied:
Instead of just checking for the error being `nil`, what I think we can do here is pass a custom error variable to the mock (e.g., `errors.New("failed")`) and check against this very error. It shouldn't matter what error value we inject since our code doesn't depend on any specific error -- we just want to verify that we return the error to the caller that the (mocked) API gives us.
Similar for the endpoint case.

I hope this addresses your third question.

Regner replied:
That should work great. Will make the change.

timoreimann commented on provider/kubernetes_test.go:
```diff
@@ -1543,6 +1612,10 @@ func (c clientMock) GetIngresses(namespaces k8s.Namespaces) []*v1beta1.Ingress {
 
 func (c clientMock) GetService(namespace, name string) (*v1.Service, bool, error) {
+	if c.apiServiceError {
+		return &v1.Service{}, true, c.err
```
Conventionally, if an error occurs in a function returning multiple values, the non-`error` parameters are set to their default values. Thus, I'd use `false` for the boolean return value.

Slightly related/unrelated: I don't know why we return `true` in line 1624 when we seemingly can't find the service.

timoreimann commented on provider/kubernetes_test.go:
```diff
@@ -1552,6 +1625,10 @@ func (c clientMock) GetService(namespace, name string) (*v1.Service, bool, error
 
 func (c clientMock) GetEndpoints(namespace, name string) (*v1.Endpoints, bool, error) {
+	if c.apiEndpointsError {
+		return &v1.Endpoints{}, true, c.err
```
Also `false` here.

timoreimann commented on provider/kubernetes_test.go:
```diff
@@ -1524,11 +1524,80 @@ func TestServiceAnnotations(t *testing.T) {
+			ObjectMeta: v1.ObjectMeta{
+				Namespace: "testing",
+				Annotations: map[string]string{
+					"traefik.frontend.passHostHeader": "true",
```
Do we need this annotation for the tests at hand?

I'm always in favor of keeping test fixtures as minimized as possible.

Regner replied:
Going to move it to a separate test specifically for checking the !exists cases.

timoreimann commented on provider/kubernetes_test.go:
```diff
@@ -1524,11 +1524,80 @@ func TestServiceAnnotations(t *testing.T) {
+			Spec: v1beta1.IngressSpec{
+				Rules: []v1beta1.IngressRule{
+					{
+						Host: "foo",
```
Needed?

Regner replied:
See above.

timoreimann commented on provider/kubernetes.go:
```diff
@@ -169,10 +169,13 @@ func (provider *Kubernetes) loadIngresses(k8sClient k8s.Client) (*types.Configur
+					return nil, err
+				}
+
+				if !exists {
```
I don't see a test for this case. I suppose it's already covered by a pre-existing test?

Regner replied:
You mean the does not exist part? No that was on my todo list, it is actually why the ingress part of the new test has sample data. Was thinking of moving it to a separate test though.

timoreimann replied:
Separate test sounds good. :+1:

Regner replied:
Added tests for this and the !exists endpoints

timoreimann commented on provider/kubernetes.go:
```diff
@@ -196,10 +199,10 @@ func (provider *Kubernetes) loadIngresses(k8sClient k8s.Client) (*types.Configur
 						endpoints, exists, err := k8sClient.GetEndpoints(service.ObjectMeta.Namespace, service.ObjectMeta.Name)
 						if err != nil || !exists {
```
Any particular reason we not splitting up the two cases here as well?

Regner replied:
A bad rebase is why actually. Specifically wrote some test code for it. Will fix,

timoreimann commented on provider/kubernetes_test.go:
```diff
@@ -1524,11 +1524,80 @@ func TestServiceAnnotations(t *testing.T) {
 	endpoints []*v1.Endpoints
 	watchChan chan interface{}
+
+	err               error
```
You can get rid of this variable by changing the type of `api{Service,Endpoints}Error` to `error` and returning the error if it's non-nil in the stubbed getter methods.

Regner commented:
Tests SHOULD be passing now. Still need to move the tests over to using the table driven test method, need to add tests for the !exists code paths, and am not sure how we want to handle when things don't exist. For example when a service doesn't exist do we still add the frontend? The two cases of this right now are when services or endpoints don't exist. Both cases can easily be someone putting a typo in the service name or port name.

timoreimann commented:
AFAICS, the only alternative we have to skipping frontends with missing services/endpoints is to revert the entire configuration (i.e., apply the error case resolution). IMHO, we shouldn't do that since such misses are usually due to a user mistake (as you said) and would consequently "punish" all other frontends that did everything correct.
There's also the case where a service or endpoint isn't available because the k8s reconciliation loops haven't completed yet. That's another situation where skipping and hoping for the next watch API event to trigger a benevolent update seems like the better option to me.

I believe the focus of this PR should be to fix #1240, which is to protect against the Kube API temporarily being unavailable. By handling that case gracefully (through reverting to the previous configuration) and maintaining the "missing" behavior as-is, I think we have satisfied the goal at hand. Potential improvements to the missing case, especially if the direction isn't super clear, can be handled by follow-up issues/PRs.

Just my 2 cents. :)

Regner commented:
Was reviewing this PR this morning and noticed that the condition in my new test to check for failure was actually wrong and would be passing even if it shouldn't.

Fixed that and added test data back in. Without the test data the loadIngress logic doesn't get far enough along to trigger the errors we want to test for. So added the test data back in.

Very sorry about that! The logic in loadIngress was fine and I didn't have to make any changes there. It was just the tests not actually doing anything useful.

timoreimann commented on provider/kubernetes_test.go:
```diff
@@ -1524,11 +1525,89 @@ func TestServiceAnnotations(t *testing.T) {
+	provider := Kubernetes{}
+	_, err := provider.loadIngresses(client)
+
+	if err == nil || err.Error() != "failed service call" {
```
We can simplify the assertion by comparing error values:

```go
if err != client.apiServiceError {
  ...
}
```

Once you turn the test into a table-driven one, define and inject an error variable `var apiErr = errors.New("error")` and compare against that in each test case.

timoreimann commented on provider/kubernetes_test.go:
```diff
@@ -1524,11 +1525,89 @@ func TestServiceAnnotations(t *testing.T) {
+	provider = Kubernetes{}
+	_, err = provider.loadIngresses(client)
+
+	if err == nil || err.Error() != "failed endpoints call" {
```
Same pattern as above.

emilevauge commented on provider/kubernetes_test.go:
```diff
@@ -1524,11 +1525,97 @@ func TestServiceAnnotations(t *testing.T) {
+
+	endpoints := []*v1.Endpoints{}
+	watchChan := make(chan interface{})
+	apiErr := errors.New("failed")
```
Minor comment, but I would use something more random than _failed_ here.

Regner commented:
Commits all squashed and that final test case added.

@timoreimann be aware, as this isn't obvious with the squash, that I renamed the test from `TestServiceErrors` to `TestKubeAPIErrors` as it isn't just related to services but endpoints as well.

I also changed `apiErr := errors.New("failed")` to `apiErr := errors.New("failed kube api call")` as per the comment from @emilevauge.

Other then that I only added the `TestMissingResources` test case and nothing else should have changed.

timoreimann commented on provider/kubernetes_test.go:
```diff
@@ -1524,11 +1525,283 @@ func TestServiceAnnotations(t *testing.T) {
+					},
+				},
+				{
+					Host: "bar",
```
Suggestion: Could we change the host name to something like "host_with_missing_service" or place an according comment to indicate that the purpose of this Ingress resource test-wise is to not be found because of that reason?

It takes a bit of back of forth scrolling to understand the purpose of the individual objects this test is setting up. I think purposeful names / comments could alleviate that.

timoreimann commented on provider/kubernetes_test.go:
```diff
@@ -1524,11 +1525,283 @@ func TestServiceAnnotations(t *testing.T) {
+					},
+				},
+				{
+					Host: "tar",
```
Similarly, something that indicates this Ingress' Service's Endpoints are missing?

timoreimann commented:
The new `TestMissingResources` test is still failing according to CI.

Regner commented:
My most recent commit where I add a blank Servers attribute the "missing_service" expected output is required to pass the diff. @timoreimann helped me figure this out, but essentially without it we end up comparing a nil to a blank Servers definition.

```
-          (string) (len=15) "missing_service": (*types.Backend)(0xc4201b0450)({
-           Servers: (map[string]types.Server) <nil>,
+          (string) (len=15) "missing_service": (*types.Backend)(0xc4201b0270)({
+           Servers: (map[string]types.Server) {
+           },
```

One of the unfortunate parts of that is the normal logging we do for the failed comparison doesn't show this difference. The above output was retrieved when @timoreimann made the following change:

```
if !reflect.DeepEqual(actual, expected) {
    t.Fatalf("expected %+v, got %+v", string(expectedJSON), string(actualJSON))
    t.Fatalf("expected\n%v\ngot\n\n%v", spew.Sdump(expected), spew.Sdump(actual))
}
```

Something to keep in mind in the future.

timoreimann commented:
@Regner I pushed a small commit that includes a link to #1307.

The tests went green before, so I think as soon as we have @emilevauge's approval, we can squash and merge. 🎉

Regner commented:
Good call on adding the link
